hw/drivers/pwm/pwm_stm32/src/pwm_stm32.c

```diff
@@ -0,0 +1,626 @@
+          break;
+#endif
+#ifdef TIM3
+      case (uintptr_t)TIM3:
+          LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_TIM3);
+          dev->pwm_chan_count = 4;
+          break;
+#endif
+#ifdef TIM4
+      case (uintptr_t)TIM4:
+          LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_TIM4);
+          dev->pwm_chan_count = 4;
+          break;
+#endif
+#ifdef TIM5
+      case (uintptr_t)TIM5:
+          LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_TIM5);
+          dev->pwm_chan_count = 4;
+          break;
+#endif
+
+      /* basic timers TIM6 and TIM7 have no PWM capabilities */
+
+#ifdef TIM8
+      case (uintptr_t)TIM8:
+          LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_TIM8);
+          dev->pwm_chan_count = 4;
+          break;
+#endif
+#ifdef TIM9
+      case (uintptr_t)TIM9:
+          LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_TIM9);
+          dev->pwm_chan_count = 2;
+          break;
+#endif
+#ifdef TIM10
+      case (uintptr_t)TIM10:
+          LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_TIM10);
+          dev->pwm_chan_count = 1;
+          break;
+#endif
+#ifdef TIM11
+      case (uintptr_t)TIM11:
+          LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_TIM11);
+          dev->pwm_chan_count = 1;
+          break;
+#endif
+#ifdef TIM12
+      case (uintptr_t)TIM12:
+          LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_TIM12);
+          dev->pwm_chan_count = 2;
+          break;
+#endif
+#ifdef TIM13
+      case (uintptr_t)TIM13:
+          LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_TIM13);
+          dev->pwm_chan_count = 1;
+          break;
+#endif
+#ifdef TIM14
+      case (uintptr_t)TIM14:
+          LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_TIM14);
+          dev->pwm_chan_count = 1;
+          break;
+#endif
+#ifdef TIM15
+      case (uintptr_t)TIM15:
+          LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_TIM15);
+          dev->pwm_chan_count = 2;
+          break;
+#endif
+#ifdef TIM16
+      case (uintptr_t)TIM16:
+          LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_TIM16);
+          dev->pwm_chan_count = 1;
+          break;
+#endif
+#ifdef TIM17
+      case (uintptr_t)TIM17:
+          LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_TIM17);
+          dev->pwm_chan_count = 1;
+          break;
+#endif
+
+      /* basic timer TIM18 has no PWM capabilities */
+
+#ifdef TIM19
+      case (uintptr_t)TIM19:
+          LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_TIM19);
+          dev->pwm_chan_count = 4;
+          break;
+#endif
+#ifdef TIM20
+      case (uintptr_t)TIM20:
+          LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_TIM20);
+          dev->pwm_chan_count = 4;
+          break;
+#endif
+
+      default:
+          assert(0);
+    }
+
+    for (int i=0; i < STM32_PWM_CH_MAX; ++i) {
+        stm32_pwm_ch_set_compare(pwm->timx, i, STM32_PWM_CH_IDLE);
+        pwm->pin[i] = MCU_AFIO_PIN_NONE;
+    }
+
+    dev->pwm_funcs.pwm_configure_channel = stm32_pwm_ch_confgigure;
```

Comment:
There's also the naming convention set by the API interface, why not keep it consistent?

Reply:
The api doesn't make the distinction between device and channel functions which I find is quite a loss in expressiveness.

Reply:
How does pwm_configure_channel vs pwm_configure_device not make distinction between device and channel functions?

Reply:
The questions being addressed are
* what's the difference between `pwm_set_frequency` and `pwm_set_duty_cycle` ?
* what do `pwm_configure_channel` and `pwm_set_duty_cycle` have in common?
and more general
* which api calls modify/affect the device - and there by all channels of it?
* which api calls are specific to a channel and have no impact on any other channel?

Reply:
@mlampert I see where you are going with this line of rhetorical questioning. Although let me remember you that, a few weeks ago, [I asked you about your opinion](https://github.com/apache/mynewt-core/issues/955) on possible changes on the API _before_ I changed it and refactored two drivers and you didn't mention this. Despite this constant requesting for new changes on the API and the intent to diverge from a naming convention clearly suggested on the API, I will go through the process of explaining why we aren't using suffixes specifying whether *all* actions affect a single channel or the whole device:
* `pwm_configure_device` and `pwm_configure_channel` configure two different things, so, same action, different objects. We use the suffixes for disambiguation.
* `pwm_set_frequency` sets a frequency for a device, It can't be otherwise, since all channels on a device must run at the same frequency.
* `pwm_set_duty_cycle` sets the duty cycle value for a given channel, since this method allows the user to specify a channel number, it is kind of obvious it shouldn't affect other channels.
* If we take into consideration that a PWM device is typically a clock divider we know that every channel will work at the same frequency and, since there are multiple channels per device it makes a lot of sense that the duty cycle is set to a channel.
* Another helpful asset here is the documentation, even though assuming the user knows what PWM devices are(or just look at the methods' signature).